synapse/visibility.py

```diff
@@ -405,36 +439,26 @@ def check_event_is_visible(event: EventBase, state: StateMap[EventBase]) -> bool
         for key, event_id in key_to_eid.items()
     }
 
-    def include(typ, state_key):
-        if typ != EventTypes.Member:
-            return True
-
+    def include(state_key):
         # we avoid using get_domain_from_id here for efficiency.
         idx = state_key.find(":")
         if idx == -1:
             return False
         return state_key[idx + 1 :] == server_name
 
     event_map = await storage.main.get_events(
-        [e_id for e_id, key in event_id_to_state_key.items() if include(key[0], key[1])]
+        [
+            e_id
+            for e_id, (_, state_key) in event_id_to_state_key.items()
+            if include(state_key)
+        ]
```

Comment:
Sanity check: for this to be safe, event_id_to_state_key should only contain membership events. That's true by construction on +426.